### PR TITLE
[Feature](variant) add variant datatype related metadata

### PR DESCRIPTION
**Pull request by eldenmoon**
## Proposed changes

Add variant type for metadata Add persistent information for variant, including the path of variant sub-columns, persisting them to the segment footer and tablet schema of the rowset.

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

**Comment by eldenmoon** [comment]
run buildall

File: gensrc/thrift/Descriptors.thrift [hunk 1]
```diff
@@ -60,6 +60,8 @@ struct TSlotDescriptor {
   // materialize them.Used to optmize to read less data and less memory usage
   13: optional bool need_materialize = true
   14: optional bool is_auto_increment = false;
+  // `$.a.b.c` => ['$', 'a', 'b', 'c']
```
Comment (xiaokang) [inline gensrc/thrift/Descriptors.thrift:63]: correct comment

File: gensrc/proto/segment_v2.proto [hunk 2]
```diff
@@ -162,6 +184,12 @@ message ColumnMetaPB {
     optional uint64 num_rows = 11;
     repeated string children_column_names = 12;
 
+    // persist info for PathInData that represents path in document, e.g. JSON.
+    optional ColumnPathInfo column_path_info = 13;
+
+    optional bytes default_value = 14; // ColumnMessage.default_value ?
```
Comment (xiaokang) [inline gensrc/proto/segment_v2.proto:190]: add comment to explain why add them

File: gensrc/proto/segment_v2.proto [hunk 3]
```diff
@@ -162,6 +184,12 @@ message ColumnMetaPB {
     optional uint64 num_rows = 11;
     repeated string children_column_names = 12;
 
+    // persist info for PathInData that represents path in document, e.g. JSON.
```
Comment (xiaokang) [inline gensrc/proto/segment_v2.proto:187]: add comment to explain each extracted subcolumn of variant column will generate a column in be schema

File: gensrc/proto/olap_file.proto [hunk 4]
```diff
@@ -206,6 +206,7 @@ message ColumnPB {
     repeated ColumnPB children_columns = 17;
     repeated string children_column_names = 18;
     optional bool result_is_nullable = 19;
+    optional segment_v2.ColumnPathInfo column_path_info = 20;
```
Comment (xiaokang) [inline gensrc/proto/olap_file.proto:209]: add comment to explain each extracted subcolumn of variant column will generate a column in be schema

**Review by xiaokang — COMMENTED** [review]
add some comment for storage meta changes

File: gensrc/proto/segment_v2.proto (hunk 3 above)
Comment (eldenmoon) [reply gensrc/proto/segment_v2.proto:187]: comment already in ColumnPathInfo

File: gensrc/proto/segment_v2.proto (hunk 2 above)
Comment (eldenmoon) [reply gensrc/proto/segment_v2.proto:190]: done

File: gensrc/proto/olap_file.proto (hunk 4 above)
Comment (eldenmoon) [reply gensrc/proto/olap_file.proto:209]: comment already in ColumnPathInfo

**Comment by eldenmoon** [comment]
run buildall

File: gensrc/thrift/Descriptors.thrift (hunk 1 above)
Comment (eldenmoon) [reply gensrc/thrift/Descriptors.thrift]: done

**Comment by eldenmoon** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Pull request merged**